### DISCUSSION TITLE
Adding WebSerial - big job?

### DISCUSSION THREAD

**Discussion by YouCanNotBeSerious** [HomeSpan/HomeSpan#743] [Q&A] [answered]
I found this https://randomnerdtutorials.com/esp32-webserial-library/ and thought WebSerial would be useful for several of my HomeSpan implementations, for devices that are remote.

I added webserial in to the simple light bulb example (number 1, about as simple as it gets!) but the ESP simply reboots continually with:

```
11:48:41.061 -> Device Name:      HomeSpan LightBulb
11:48:41.061 -> 
11:48:46.072 -> 
11:48:46.072 -> assert failed: tcpip_api_call IDF/components/lwip/lwip/src/api/tcpip.c:497 (Invalid mbox)
```

**Simple question - am I chasing a big problem here or should this work and I just need to start troubleshooting?**

This might be related to https://github.com/HomeSpan/HomeSpan/discussions/88 which seems to suggest maybe there is an underlying reason WebSerial won't easily work, in that something other than AsyncTCP is already in use?

I am not starting the webserial instance on port 80 since HomeSpan already needs that - i tried 8080. It looks like the error/crash is to do with TCP sockets, googling suggests if you try and start things before underlying TCP things are ready it might happen, that's as far as I have got.

**Comment by HomeSpan** [comment]
I believe others have used Web Serial but it's not something that I'm planning on supporting in HomeSpan - not sure what the issues may be.

**Reply by YouCanNotBeSerious** [reply, marked as answer]
**PROGRESS....**  the scary complexity of the error message and subsequent dump made me think this would be beyond my ability.  

Turns out webserial was trying to start before wifi was up.... seems like a nice simple error to that effect would help :-). I guess there is an indicated error like that because a lot of people have found similar with various code trying to start too early....

Anyway, I wait for homespan to start up then start webserial and it's all fine!

Next i will try for a way to know that wifi is ready or homespan is finished starting, instead of just a dumb 5 second delay that i have now. 

NOTE for anyone that wants to try this, remember not to use port 80 for webserial.

**Reply by HomeSpan** [reply]
Note that HomeSpan includes a variety of callbacks, including one that gets called when WiFi first connects, which should solve your issue.  Check out the API for details.

In general, I recommend that anyone who is looking to extend HomeSpan start by skimming through the whole API just to get a good idea of all the different functions that already exist that could help.  I often do that with any new module I encounter - sometimes even reading through the include files gives a good sense of what's possible.

**Reply by frankonski** [reply]
Make sure to stop your webserver / webserial on wifi disconnect, and restart it on reconnect.

**Reply by YouCanNotBeSerious** [reply]
Thanks for the note - but how exactly?  I have the wificallback **starting** it fine but the reference for that function doesn't mention a related stop function ... I guess I could check occasionally with   ```while (WiFi.status() != WL_CONNECTED) {```

**Reply by frankonski** [reply]
HS_STATUS will help you. https://github.com/HomeSpan/HomeSpan/blob/master/docs/HS_STATUS.md

Use the callback. On wifi connect: start webserial. On wifi disconnect: stop webserial.

**Reply by frankonski** [reply]
My bad, I thought it had `HS_WIFI_CONNECTED` and `HS_WIFI_DISCONNECTED` ... but it doesn't (for now).

**Reply by HomeSpan** [reply]
FYI,

After the release (hopefully soon) of 1.8.1 I plan on updating the connectivity logic to allow for more flexibility to support ethernet-based devices as well as enterprise-WPA.  To do so I will likely adopt the Espressif event handling for connectivity, which is distinct from the Espressif event handling for web servers (and which I don't  be using for HAP itself)

**Reply by frankonski** [reply]
No offence (this is actually a compliment!) but there is **a lot** of good new stuff in 1.8.1-dev to call it a _patch_ on top of 1.8. I'd call it 1.9 ;-)

**Reply by HomeSpan** [reply]
Yeah, that probably makes sense.  I'll bump to 1.9.0, especially since it has been a while since last update.

**Comment by frankonski** [comment]
That is relatively similar to what I intended to do with logging callbacks.
Browser hits HomeSpan “custom” weblogs, http handler keeps the connection open with keepalive, logging callback pushes logs back down the http connection to the browser, who refreshes real time…

real-time weblogs…

missing time to work on this.

**Reply by YouCanNotBeSerious** [reply]
@frankonski "real time weblogs" is basically what I am trying to achieve :-) 

I am continuing to mess with webserial as it seems like an easy first step - works well on other projects I have but something in Homespan is interfering, or rather i should say webserial interferes with something in homespan as homespan works as designed!
